### ISSUE TITLE
AudioWorkletNode is abstract and has no constructor

### ISSUE THREAD

**Issue by Larpoux** [fzyzcjy/flutter_rust_bridge#2193]
### Describe the bug

AudioWorkletNode is defined in Rust as : 
```rust
/// A user-defined AudioNode which lives in the control thread
///
/// - MDN documentation: <https://developer.mozilla.org/en-US/docs/Web/API/AudioWorkletNode>
/// - specification: <https://webaudio.github.io/web-audio-api/#AudioWorkletNode>
///
/// # Examples
///
/// - `cargo run --release --example worklet`
/// - `cargo run --release --example worklet_message_port`
/// - `cargo run --release --example worklet_bitcrusher`
///
#[derive(Debug)]
pub struct AudioWorkletNode {
    registration: AudioContextRegistration,
    channel_config: ChannelConfig,
    number_of_inputs: usize,
    number_of_outputs: usize,
    audio_param_map: HashMap<String, AudioParam>,
}

impl AudioWorkletNode {
    /// Construct a new AudioWorkletNode
    ///
    /// # Panics
    ///
    /// This function panics when
    /// - the number of inputs and the number of outputs of the supplied options are both equal to
    /// zero.
    /// - any of the output channel counts is equal to zero or larger than 32 ([`MAX_CHANNELS`])
    pub fn new<P: AudioWorkletProcessor + 'static>(
        context: &impl BaseAudioContext,
        options: AudioWorkletNodeOptions<P::ProcessorOptions>,
    ) -> Self {
         ...
}
```
The constructor is missing in the dart file generated : 
```dart
// These functions are ignored because they have generic arguments: `new`, `process`
// These types are ignored because they are not used by any `pub` functions: `AudioParamValues`, `AudioWorkletNodeOptions`, `AudioWorkletRenderer`, `Processor`
// These function are ignored because they are on traits that is not defined in current crate (put an empty `#[frb]` on it to unignore): `clone`, `fmt`, `fmt`, `fmt`, `has_side_effects`, `onmessage`, `process`
// These functions are ignored (category: IgnoreBecauseExplicitAttribute): `constructor`, `onmessage`, `port`
// These functions are ignored (category: IgnoreBecauseOwnerTyShouldIgnore): `default`

// Rust type: RustOpaqueMoi<flutter_rust_bridge::for_generated::RustAutoOpaqueInner<AudioWorkletNode>>
abstract class AudioWorkletNode implements RustOpaqueInterface, AudioNode {
   ...
}
```
Certainly because of the generic parameter `<P: AudioWorkletProcessor + 'static>`

### Steps to reproduce

-

### Logs

```shell
-
```

**Comment by fzyzcjy** [comment]
That looks like because generics is not supported yet. However, IIRC worklet requires users to provide a function to run, and it has nontrivial signatures. Thus I suggest working on it later ;)

**Comment by Larpoux** [comment]
I have also a problem with MediaRecorder. The following function is not generated : 
```rust
pub fn set_ondataavailable<F: FnMut(BlobEvent) + Send + 'static>(&self, callback: F)
```

**Comment by fzyzcjy** [comment]
For this one, try to mimic the `set_blahblah` we have dealt with many times

**Comment by Larpoux** [comment]
Yes, I think so. I don't understand why this function has a so complicated signature. I am going to do that myself, when I will have finished to port all the Mozilla examples to flutter.
Today I have ported 13 Mozilla examples. Only two of those examples was unsuccessful.
You really did a good job. 👍

**Comment by fzyzcjy** [comment]
That looks great and thank you!

**Comment by Larpoux** [comment]
Again a problem with class `MediaElement` : the constructor is missing
```rust
    pub fn new<P: Into<PathBuf>>(file: P) -> Result<Self, Box<dyn Error>> 
```
I continue not understanding why Rust Web Audio API uses so complicated functions.
The constructor just need a file (or perhaps a file name) as argument.
No reason to do complicate sig. that I can't understand because of my absence of Rust knowledge.

**Comment by fzyzcjy** [comment]
For this spefici thing, maybe 

```rust
pub fn frb_override_new(file: String) -> anyhow::Result<Self> {
  Self::new(file)
}
```

(for error handling, check other functions where we use `anyhow!` or something like that)

> I continue not understanding why Rust Web Audio API uses so complicated functions.

I guess that's usual life in Rust - we use such generics for better APIs.

**Comment by Larpoux** [comment]
In class `AudioBuffer` : 
```rust
pub fn get_channel_data(&self, channel_number: usize) -> &[f32] 
```
generates a void function : 
```dart
void getChannelData({required int channelNumber});
```
If we can't provide a pointer into the buffer, probably a function `void setAt(int index)` and a function `float getAt(int index)` would be OK.

**Comment by fzyzcjy** [comment]
When returning `&[f32]`, it seems it will not be mutable, so we do not need to provide any `set` functions.

Btw IIRC we have seen such "return a `&[Something]`", thus maybe we can refer to it, e.g. by making it a `Vec<Something>`

**Comment by Larpoux** [comment]
In override_web_audio_api_rs:
```rust
#[ext(name = MediaElementMiscExt)]
pub impl MediaElement {
    fn frb_override_new(file: String) -> anyhow::Result<Self>
    where
        Self: std::marker::Sized,
    {
        Self::new(file).map_err(|e| anyhow::anyhow!("{:?}", e))
    }
}
```
is not generated because:
```dart
// These functions are ignored (category: IgnoreBecauseOwnerTyShouldIgnore): `frb_override_new`
```
Please, what does it mean ?

**Comment by fzyzcjy** [comment]
Looks like MediaElement is ignored, and we made a MyMediaElement (yes we should rename the dart name and the "My" is ugly...)

![image](https://github.com/user-attachments/assets/c893b438-7529-4be1-9557-f3fd918b714b)
![image](https://github.com/user-attachments/assets/6455b593-712d-4293-8ae5-6d4613c56f0d)

**Comment by fzyzcjy** [comment]
Close since this seems to be solved, but feel free to reopen if needed!

**Comment by Larpoux** [comment]
> ### Describe the bug
> AudioWorkletNode is defined in Rust as :
> 
> ```rust
> /// A user-defined AudioNode which lives in the control thread
> ///
> /// - MDN documentation: <https://developer.mozilla.org/en-US/docs/Web/API/AudioWorkletNode>
> /// - specification: <https://webaudio.github.io/web-audio-api/#AudioWorkletNode>
> ///
> /// # Examples
> ///
> /// - `cargo run --release --example worklet`
> /// - `cargo run --release --example worklet_message_port`
> /// - `cargo run --release --example worklet_bitcrusher`
> ///
> #[derive(Debug)]
> pub struct AudioWorkletNode {
>     registration: AudioContextRegistration,
>     channel_config: ChannelConfig,
>     number_of_inputs: usize,
>     number_of_outputs: usize,
>     audio_param_map: HashMap<String, AudioParam>,
> }
> 
> impl AudioWorkletNode {
>     /// Construct a new AudioWorkletNode
>     ///
>     /// # Panics
>     ///
>     /// This function panics when
>     /// - the number of inputs and the number of outputs of the supplied options are both equal to
>     /// zero.
>     /// - any of the output channel counts is equal to zero or larger than 32 ([`MAX_CHANNELS`])
>     pub fn new<P: AudioWorkletProcessor + 'static>(
>         context: &impl BaseAudioContext,
>         options: AudioWorkletNodeOptions<P::ProcessorOptions>,
>     ) -> Self {
>          ...
> }
> ```
> 
> The constructor is missing in the dart file generated :
> 
> ```dart
> // These functions are ignored because they have generic arguments: `new`, `process`
> // These types are ignored because they are not used by any `pub` functions: `AudioParamValues`, `AudioWorkletNodeOptions`, `AudioWorkletRenderer`, `Processor`
> // These function are ignored because they are on traits that is not defined in current crate (put an empty `#[frb]` on it to unignore): `clone`, `fmt`, `fmt`, `fmt`, `has_side_effects`, `onmessage`, `process`
> // These functions are ignored (category: IgnoreBecauseExplicitAttribute): `constructor`, `onmessage`, `port`
> // These functions are ignored (category: IgnoreBecauseOwnerTyShouldIgnore): `default`
> 
> // Rust type: RustOpaqueMoi<flutter_rust_bridge::for_generated::RustAutoOpaqueInner<AudioWorkletNode>>
> abstract class AudioWorkletNode implements RustOpaqueInterface, AudioNode {
>    ...
> }
> ```
> 
> Certainly because of the generic parameter `<P: AudioWorkletProcessor + 'static>`




> That looks like because generics is not supported yet, so try to avoid the generics. However, IIRC worklet requires users to provide a function to run, and it has nontrivial signatures (iirc something like `&mut [&mut [&mut [u8]]]`). Thus I suggest working on it later ;)


No, this is not fixed. And `AudioWorkletNode` is a very important node for me.
Do you have an idea how to fix that ?
Is there any hope to have this supported by FRB in a foreseeable future ?

**Comment by fzyzcjy** [comment]
Oops, I only checked the last few comments and thought that one was solved.

> Do you have an idea how to fix that ?
Is there any hope to have this supported by FRB in a foreseeable future ?

As mentioned earelier:

> However, IIRC worklet requires users to provide a function to run, and it has nontrivial signatures (iirc something like &mut [&mut [&mut [u8]]]). Thus I suggest working on it later ;)

If you do not care about performance much, then it can be implemented today: Just write wrapper classes/methods that converts such complex types to what frb understands today (e.g. `Vec<Vec<Vec<u8>>>` or whatever).
